tests/unit/frameworks/pytorch/lr_scheduler_test.py

```diff
@@ -53,5 +53,153 @@ def test_chaining_schedulers(self):
         self.assertEqual(expected, actual)
 
 
+class ComposedLRSchedulerTest(unittest.TestCase):
+    def test_epoch_update(self):
+        optimizer = torch.optim.SGD([torch.zeros(1)], lr=1.0)
+
+        # StepLR(optimizer, step_size=2, gamma=0.1) for first 10 epochs
+        expected = np.repeat([1.0, 0.1, 0.01, 0.001, 0.0001], 2).tolist()
+
+        # StepLR(optimizer, step_size=3, gamma=2) for last 15 epochs
+        expected += np.repeat([1.0, 2.0, 4.0, 8.0, 16.0], 3).tolist()
+
+        scheduler = ComposedLRScheduler(
+            optimizer,
+            schedulers={
+                0: dict(
+                    lr_scheduler_class=StepLR,
+                    lr_scheduler_args=dict(step_size=2, gamma=0.1),
+                ),
+                10: dict(
+                    lr_scheduler_class=StepLR,
+                    lr_scheduler_args=dict(step_size=3, gamma=2),
+                ),
+            },
+        )
+        actual = []
+        for _ in range(len(expected)):
+            lr = optimizer.param_groups[0]["lr"]
+            actual.append(round(lr, 4))
+            # 10 batches
+            for _ in range(10):
+                optimizer.step()
+            # Called once per epoch
+            scheduler.step()
+
+        self.assertEqual(expected, actual)
+
+    def test_batch_update(self):
+
+        # OneCycleLR(optimizer, epochs=10, max_lr=2.0, steps_per_epoch=10)
+        # for first 10 epochs
+        expected = [
+            0.08,
+            0.5903,
+            1.5787,
+            1.999,
+            1.8806,
+            1.5878,
+            1.1786,
+            0.734,
+            0.3421,
+            0.0805,
+        ]
+        # StepLR(optimizer, step_size=3, gamma=0.1) for next 10 epochs
+        expected += np.repeat([1.0, 0.1, 0.01, 0.001, 0.0001], 2).tolist()
+
+        # OneCycleLR(optimizer, epochs=10, max_lr=2.0, steps_per_epoch=10)
+        # for last 10 epochs
+        expected += [
+            0.08,
+            0.5903,
+            1.5787,
+            1.999,
+            1.8806,
+            1.5878,
+            1.1786,
+            0.734,
+            0.3421,
+            0.0805,
+        ]
+
+        optimizer = torch.optim.SGD([torch.zeros(1)], lr=1.0)
+        batch_size = 10
```

Comment:
I'm confused. Why is steps_per_epoch the same as the batch_size? For clarity it might be better to call this variable `num_batches_per_epoch` or something like that.

Reply:
In the context of learning rate schedulers this parameter represent how many times the method `step` will be called per epoch which can be `number of batches` in the case of `OneCycleLR` or `1` for other learning rate schedulers that calls `step` only once per epoch.
`OneCycleLR` also calls this parameter `steps_per_epoch` for the same reason.